### ISSUE TITLE
[Poll] Support higher-order gradient computation in DDP

### ISSUE THREAD

**Issue by rohan-varma** [pytorch/pytorch#63929]
## 🚀 Feature
Computing higher-order gradients has several use cases, i.e. applying a local gradient penalty before running the backwards pass. However, DDP doesn't have great support for higher-order gradients as evidenced by https://github.com/pytorch/pytorch/issues/63812. A couple of issues:
1. `.grad` field of the gradient is not synchronized across processes
2. Calling `.backward()` twice is not well supported: https://github.com/pytorch/pytorch/issues/63812#issuecomment-904939795, as it will cause issues in DDP's internal state
3. Leaf grads accumulated by DDP don't have proper autograd recording, i.e. the allreduce is unknown to autograd and therefore we cannot differentiate through it.

If your use case requires DDP to work with higher-order gradients, please ping this issue with some details about your use case. Thanks!

**Comment by choltz95** [comment]
Hi all! Appreciate the paper trail. Just recording with my intended usecase: interested in integrating DDP with meta learning/maml algorithms. Necessitates ```autograd.grad``` with ```create_graph.```

Curious if anyone has found/has ideas any possible hacky solutions or workarounds. I guess for now I will revert to DP.

**Comment by rohan-varma** [comment]
cc @cbalioglu for potential DP use case that DDP doesn't support.

@choltz95 Just for complete understanding, are you able to provide a minimal repro under which your training fails? Thank yoU!

**Comment by zhiyuanpeng** [comment]
> Hi all! Appreciate the paper trail. Just recording with my intended usecase: interested in integrating DDP with meta learning/maml algorithms. Necessitates `autograd.grad` with `create_graph`.
> 
> Curious if anyone has found/has ideas any possible hacky solutions or workarounds. I guess for now I will revert to DP.

I have tried DP in Pytorch lightning, compared with DDP, DP performes a little worse. I have the same issue with DDP when work on Meta Leanring. Hope the ddp can support `autograd.grad`

**Comment by ppwwyyxx** [comment]
> Computing higher-order gradients has several use cases, i.e. applying a local gradient penalty before running the backwards pass

It appears that the gradient penalty case already works @rohan-varma . I ran something like this:
```python
class MyModel(nn.Module):
    def __init__(self, pre, m):
        super().__init__()
        self.pre = pre
        self.m = m

    def forward(self, x):
        x = self.pre(x)
        out = self.model(x)

        grad = torch.autograd.grad(outputs=out, inputs=x,
                                   grad_outputs=torch.ones_like(out), retain_graph=True,
                                   create_graph=True)[0]

        loss = self.model(x)
        return loss.mean(dim=0).sum() + grad.mean(dim=0).sum()   # the total loss
```
In the above example I added gradient penalty to the loss. I then wrap MyModel with DDP and call backward() on its output loss.
In a few small tests I run it appears that this gives me the right gradients. However, given this issue I want to ask if it is actually supposed to work / are there problems we should be aware of.

**Comment by jiayixu64** [comment]
Can you help prioritize this feature? We are implementing a new training optimization technique that uses this feature to compute second order gradients.

**Comment by fordflip** [comment]
bump, would also appreciate this

**Comment by hnanhtuan** [comment]
Any updates on this features? Hope to have it soon.

**Comment by xuxingc** [comment]
Hi experts, I am looking forward to this amazing feature, which can be directly applied to distributed training of meta learning, hyperparameter optimization, bi-level optimization, etc. Thanks!
